### PR TITLE
UNet Implementation

### PR DESCRIPTION
**Pull request by RaphaelMeudec**
Here is an implementation of [U-Net: Convolutional Networks for Biomedical Image Segmentation
](https://arxiv.org/abs/1505.04597).

Some remarks on the implementation:

- I didn't used the `exclude_top` option as it doesn't seem to fit with the network's architecture
- I added a `padding` option to either `same` or `valid` to allow users to either follow the original paper (`valid`) or keep the same image size between input and output (`same`)
- Training is running right now on the [ISBI Challenge](http://brainiac2.mit.edu/isbi_challenge/) and is using [this code](https://github.com/raphaelmeudec/unet-training)

### CODE REVIEW

**Comment by RaphaelMeudec** [comment]
I'm not sure on what the tests should check. For classification, the check is to have the elephant on the top 3. Do you have any clue on performance test for segmentation?

**Comment by taehoonlee** [comment]
@RaphaelMeudec, LGTM, thank you for the PR. Here are my replies:

- I didn't used the `exclude_top` option as it doesn't seem to fit with the network's architecture
  - OK.
- I added a `padding` option to either `same` or `valid` to allow users to either follow the original paper (`valid`) or keep the same image size between input and output (`same`)
  - OK.
- Training is running right now on the ISBI Challenge and is using this code
  - I think it is better to translate [the pre-trained weights](https://lmb.informatik.uni-freiburg.de/people/ronneber/u-net/u-net-release-2015-10-02.tar.gz) (`phseg_v5.caffemodel`) from [the original repository](https://lmb.informatik.uni-freiburg.de/people/ronneber/u-net/).
- I'm not sure on what the tests should check
  - How about the output map of the sample data (`PhC-C2DH-U373/01/t000.tif`) in the author's zip file?

File: keras_applications/unet.py [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+    Optionally loads weights pre-trained on Biomedical Image segmentation dataset ?
+
+    # Arguments
+        input_tensor: optional Keras tensor (i.e. output of `layers.Input()`)
+            to use as image input for the model.
+        input_shape: optional shape tuple, only to be specified
+            if `include_top` is False (otherwise the input shape
+            has to be `(224, 224, 3)` (with `channels_last` data format)
+            or `(3, 224, 224)` (with `channels_first` data format).
+            It should have exactly 3 inputs channels,
```
Comment (gatagat) [inline keras_applications/unet.py]: Where is `exactly 3 input channels` required? The original method would work on any number of channels.

Comment (RaphaelMeudec) [reply keras_applications/unet.py]: Sorry, I copied the docstring from other architectures and didn't spot this. Indeed, the architecture can take less than 3 channels, I'll impact this soon

**Comment by hadim** [comment]
Any chance to see this merged?

**Comment by taehoonlee** [comment]
@RaphaelMeudec, How are you doing with the PR? Do you have plans to update?

**Comment by slavakurilyak** [comment]
Any updates on this?
